Userland/Libraries/LibSQL/Key.cpp

```diff
@@ -0,0 +1,204 @@
+        auto& part = m_part_definitions[ix];
+        if (part.name() == n) {
+            return (int)ix;
+        }
+    }
+    return -1;
+}
+
+void Key::append(const Value& value)
+{
+    VERIFY(m_part_definitions.size() == 0);
+    m_key_data.append(value);
+}
+
+Key& Key::operator+=(Value const& value)
+{
+    append(value);
+    return *this;
+}
+
+bool Key::is_compatible(Key const& other) const
+{
+    if ((m_part_definitions.size() == 0) && (other.m_part_definitions.size() == 0)) {
+        return true;
+    }
+    if (m_part_definitions.size() != other.m_part_definitions.size()) {
+        return false;
+    }
+    for (auto ix = 0u; ix < m_part_definitions.size(); ix++) {
+        auto& my_part = m_part_definitions[ix];
+        auto& other_part = other.m_part_definitions[ix];
+        if (my_part.type() != other_part.type()) {
+            return false;
+        }
+        if (my_part.sort_order() != other_part.sort_order()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+String Key::to_string() const
+{
+    StringBuilder builder;
+    for (auto& part : m_key_data) {
+        if (!builder.is_empty()) {
+            builder.append('|');
+        }
+        auto str_opt = part.to_string();
+        builder.append((str_opt.has_value()) ? str_opt.value() : "(null)");
+    }
+    if (pointer() != 0) {
+        builder.appendff(":{}", pointer());
+    }
+    return builder.build();
+}
+
+size_t Key::size() const
+{
+    size_t sz = sizeof(u32);
+    for (auto& part : m_key_data) {
+        sz += part.size();
+    }
+    return sz;
+}
+
+void Key::copy_from(const Key& other)
+{
+    m_part_definitions.clear();
+    for (auto& part : other.m_part_definitions) {
+        m_part_definitions.append(part);
+    }
+    m_key_data.clear();
+    for (auto& part : other.m_key_data) {
+        m_key_data.append(part);
+    }
+    m_pointer = other.pointer();
+}
+
+int Key::compare(const Key& other) const
+{
+    auto ret = 0;
+    auto num_values = min(m_key_data.size(), other.m_key_data.size());
+    VERIFY(num_values > 0);
+    for (auto ix = 0u; ix < num_values; ix++) {
+        ret = m_key_data[ix].compare(other.m_key_data[ix]);
+        if (ret != 0) {
+            return ((ix < m_part_definitions.size()) && m_part_definitions[ix].sort_order() == SortOrder::Descending) ? -ret : ret;
+        }
+    }
+    return 0;
+}
+
+int Key::match(const Key& other) const
+{
+    auto ret = 0;
+    for (auto ix = 0u; ix < m_key_data.size(); ix++) {
+        if (other[ix].is_null()) {
+            return 0;
+        }
+        ret = m_key_data[ix].compare(other[ix]);
+        if (ret != 0) {
+            return (m_part_definitions[ix].sort_order() == SortOrder::Descending) ? -ret : ret;
+        }
+    }
+    return 0;
+}
+
+u32 Key::hash() const
+{
+    u32 ret = 0u;
+    for (auto& value : m_key_data) {
+        // This is the algorithm used in Boost:
+        ret ^= value.hash() + 0x9e3779b9 + (ret << 6) + (ret >> 2);
```

Comment:
Could we use our home-grown hash functions in AK/HashFunctions.h?

Reply:
I guess I could use subsequent `pair_int_hash` call but then I'd be hashing twice- first to hash the actual value (string, int, ...) and then `pair_int_hash` will hash the hash again. Or I can lift the constants from `pair_int_hash` and use them here.

I asked on Discord about that but got crickets.